source/common/http/header_map_impl.h

```diff
@@ -46,8 +46,14 @@ class HeaderMapImpl : public HeaderMap {
   static void appendToHeader(HeaderString& header, absl::string_view data);
 
   HeaderMapImpl();
-  HeaderMapImpl(const std::initializer_list<std::pair<LowerCaseString, std::string>>& values);
+  explicit HeaderMapImpl(
+      const std::initializer_list<std::pair<LowerCaseString, std::string>>& values);
   HeaderMapImpl(const HeaderMap& rhs);
+  // The above constructor for HeaderMap is not an actual copy constructor. This also prevent the
+  // implicit move constructor, moving HeaderMapImpl is unsafe (see HeaderList comments).
+  HeaderMapImpl(const HeaderMapImpl& rhs);
+  // Safe copy assignment; this is highly inefficient, don't use this except for tests.
+  HeaderMapImpl& operator=(const HeaderMapImpl& rhs);
```

Comment:
Sorry to have more comments on this, but my concern about these changes is that we can accidentally start copying without knowing about it. Is it not possible to make HeaderMapImpl NonCopyable but then add the copy/assignment constructors in TestHeaderMapImpl?

Reply:
That's a fair point. Having copy-ctor and assign-operator illegal is fine, but it would be nice to have copyFrom() as an explicit call.

Reply:
Yes, we can do this then. I'm honestly surprised what a mess this all is. If anyone wants ammunition in the language wars that points to C++ being a time suck and unreasonably complicated for even experienced programmers to work with, this is a prime example.